source/gfoidl.Stochastics/Statistics/Sample.AverageVarianceCore.cs

```diff
@@ -79,71 +102,72 @@ private unsafe void CalculateAverageAndVarianceCoreImpl(int i, int n, out double
 
-                        arr += 4 * Vector<double>.Count;
-                        i   += 4 * Vector<double>.Count;
+                        current += 4 * Vector<double>.Count;
+                        i       += 4 * Vector<double>.Count;
                     }
 
                     m = n & ~(2 * Vector<double>.Count - 1);
                     if (i < m)
                     {
-                        Core(arr, 0 * Vector<double>.Count, ref avgVec0, ref var0, end);
-                        Core(arr, 1 * Vector<double>.Count, ref avgVec1, ref var1, end);
+                        Core(current, 0 * Vector<double>.Count, ref avgVec0, ref var0, end);
+                        Core(current, 1 * Vector<double>.Count, ref avgVec1, ref var1, end);
 
-                        arr += 2 * Vector<double>.Count;
-                        i   += 2 * Vector<double>.Count;
+                        current += 2 * Vector<double>.Count;
+                        i       += 2 * Vector<double>.Count;
                     }
 
                     m = n & ~(1 * Vector<double>.Count - 1);
                     if (i < m)
                     {
-                        Core(arr, 0 * Vector<double>.Count, ref avgVec0, ref var0, end);
+                        Core(current, 0 * Vector<double>.Count, ref avgVec0, ref var0, end);
 
-                        arr += 1 * Vector<double>.Count;
+                        current += 1 * Vector<double>.Count;
                     }
 
                     // Reduction -- https://github.com/gfoidl/Stochastics/issues/43
                     avgVec0 += avgVec1 + avgVec2 + avgVec3;
-                    tmpAvg   = avgVec0.ReduceSum();
+                    tmpAvg  += avgVec0.ReduceSum();
 
-                    var0        += var1 + var2 + var3;
-                    tmpVariance  = var0.ReduceSum();
-                }
+                    var0            += var1 + var2 + var3;
+                    tmpVariance += var0.ReduceSum();
 
-                while (arr < end)
-                {
-                    tmpAvg      += *arr;
-                    tmpVariance += *arr * *arr;
-                    arr++;
+                    if (current < end)
+                    {
+                        sequentialEnd = end;
+                        goto Sequential;            // second pass for sequential
+                    }
                 }
 
+            Exit:
                 avg      = tmpAvg;
                 variance = tmpVariance;
             }
             //-----------------------------------------------------------------
             void Core(double* arr, int offset, ref Vector<double> avgVec, ref Vector<double> var, double* end)
             {
 #if DEBUG_ASSERT
-                Debug.Assert(arr + offset < end);
+                // arr is included -> -1
+                Debug.Assert(arr + offset + Vector<double>.Count - 1 < end);
 #endif
                 Vector<double> vec = VectorHelper.GetVector(arr + offset);
```

Comment:
Add a comment that unaligned read isn't necessary, because `arr` was aligned.

Reply:
Side note: on modern cpus unaligned reads are equally fast as aligned reads are.